orttraining/orttraining/core/optimizer/sce_loss_grad_bias_fusion.cc

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/core/optimizer/sce_loss_grad_bias_fusion.h"
+
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/utils.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+
+namespace onnxruntime {
+
+Status SceLossGradBiasFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level,
+                                        const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+
+  for (auto node_index : node_topology_list) {
+    auto* node_ptr = graph.GetNode(node_index);
+    if (!node_ptr) continue;  // Node was removed.
+
+    auto& node = *node_ptr;
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "SoftmaxCrossEntropyLossInternalGrad", {1}, kMSDomain) ||
+        !graph_utils::IsSupportedProvider(node, GetCompatibleExecutionProviders()) || node.InputDefs().size() == 6 ||
+        node.GetOutputEdgesCount() != 1) {
+      continue;
+    }
+
+    NodeArg* sce_grad_def = node.MutableOutputDefs()[0];
+    Node* p_next = graph.GetNode(node.OutputNodesBegin()->Index());
+    Node* p_reshape = nullptr;
+    if (graph_utils::IsSupportedOptypeVersionAndDomain(*p_next, "Reshape", {5, 13, 14}) &&
+        graph_utils::IsSupportedProvider(*p_next, GetCompatibleExecutionProviders()) &&
+        p_next->GetOutputEdgesCount() == 1) {
+      p_reshape = p_next;
+      sce_grad_def = p_reshape->MutableOutputDefs()[0];
+      p_next = graph.GetNode(p_next->OutputNodesBegin()->Index());
+    }
+
+    if (!(graph_utils::IsSupportedOptypeVersionAndDomain(*p_next, "Add", {7, 13, 14}) ||
+          graph_utils::IsSupportedOptypeVersionAndDomain(*p_next, "Sum", {6, 8, 13})) ||
+        !graph_utils::IsSupportedProvider(*p_next, GetCompatibleExecutionProviders()) ||
+        p_next->InputDefs().size() != 2) {
+      continue;
+    }
+
+    Node& sum_node = *p_next;
+    auto& sum_input_defs = sum_node.MutableInputDefs();
+    auto shape0 = sum_input_defs[0]->Shape();
+    auto shape1 = sum_input_defs[1]->Shape();
+    if (!shape0 || !shape1 || shape0->dim_size() != shape1->dim_size()) {
+      continue;
+    }
+
+    bool has_same_shape = true;
```

Comment:
I recalled Shape can be compared with ==. 

```
>   if (input->Shape() != skip->Shape()) {
>     return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
>                            "skip is expected to have same shape as input");
>   }
```

Reply:
There are different "shape". ORT's tensor shape is onnxruntime::TensorShape, the "shape" in ONNX graph (in transformer code) is onnx::TensorShapeProto.